File: content/problems/1443-minimum-time-to-collect-all-apples-in-a-tree/solutions/apple-subtree-dfs/Solution.kt
class Solution {
    private fun dfs(node:Int,parent:Int,graph:Array<MutableList<Int>>,hasApple:List<Boolean>):Int{
        var cost=0;
        for(child in graph[node])if(child!=parent){
            val childCost=dfs(child,node,graph,hasApple);
            if(childCost>0||hasApple[child])cost+=childCost+2
        };
        return cost
    };
    fun minTime(n:Int,edges:Array<IntArray>,hasApple:List<Boolean>):Int{
        val graph=Array(n){
            mutableListOf<Int>()
        };
        for(edge in edges){
            graph[edge[0]].add(edge[1]);
            graph[edge[1]].add(edge[0])
        };
        return dfs(0,-1,graph,hasApple)
    }
}

File: content/problems/1443-minimum-time-to-collect-all-apples-in-a-tree/solutions/apple-subtree-dfs/solution.cpp
class Solution {
    int dfs(int node,int parent,vector<vector<int>>& graph,vector<bool>& hasApple){
        int cost=0;
        for(int child:graph[node])if(child!=parent){
            int childCost=dfs(child,node,graph,hasApple);
            if(childCost||hasApple[child])cost+=childCost+2;
        }
        return cost;
    }
    public:int minTime(int n,vector<vector<int>>& edges,vector<bool>& hasApple){
        vector<vector<int>> graph(n);
        for(auto& edge:edges){
            graph[edge[0]].push_back(edge[1]);
            graph[edge[1]].push_back(edge[0]);
        }
        return dfs(0,-1,graph,hasApple);
    }
};
